### ISSUE TITLE
[DI] Change the way hidden services are defined

### ISSUE THREAD

**Issue by javiereguiluz** [symfony/symfony#27284]
**Description**  
Change the way hidden services are defined.

**Example**  
Originally asked for in [this comment of the Symfony blog](https://symfony.com/blog/new-in-symfony-4-1-hidden-services#comment-22449). Should we introduce a `hidden` option for services instead of relying on the leading `.` character in the service ID?

Before:

```yaml
services:
    .App\Something\Foo: ~
```

After:

```yaml
services:
    App\Something\Foo:
        hidden: true
```

**Comment by iltar** [comment]
I'm personally :+1: and this has several reasons:
 - dot prefixing is implicit configuration
 - it's hard to describe what it is, so googling for its perpose is difficult, where as googling for "symfony hidden service" is very explicit and makes sense when you see it
 - not everyone thinks of linux files (if they even know about it) when they see it

**Comment by jenkoian** [comment]
> Should we introduce a hidden option for services **instead of** relying on the leading . character in the service ID?

How about **as well as**? I mean like either/or...

**Comment by tzfrs** [comment]
Code should be written in a way, that a person, no matter from which background he comes, can understand it. Adding a `.` at the beginning of the service definition eliminates that, because, how would he know that this makes a service hidden without knowing the documentation? Therefore I'm totally :+1: on changing it to `hidden: true`.

Random thought: This could also have the advantage, to define `hidden: true` in the `defaults` of a service definition, meaning, you COULD put all your hidden services into `services_hidden.yaml`.

**Comment by OskarStark** [comment]
@tzfrs I like your idea to put all hidden services in a `services_hidden.yaml` file adn autoconfigure them

**Comment by nicolas-grekas** [comment]
The fact that an identifier is hidden is not a semantic on the service not on the alias itself, but is a property of the identifier, i.e.e the string, not the thing that is bound to it.

It's exactly the same with hidden files on Linux: this concept doe not belong to the file content not to the file stats. It's just something that belong to the file NAME.

That's why things are this way, and there is no abuse of anything here, please don't make such statements.

**Comment by javiereguiluz** [comment]
To give some more context, to me this is similar to [hidden Symfony commands](https://symfony.com/doc/current/console/hide_commands.html). Only the command names are hidden, because you can still execute those commands ... but we use `setHidden(true)` instead of putting a `.` in the command name.

**Comment by alcaeus** [comment]
To provide some context, the idea of "hidden" files was the result of an ugly hack in the `ls` command to hide the `.` and `..` folders. There is a [nice article on hidden files](https://linux-audit.com/linux-history-how-dot-files-became-hidden-files/) with an interesting conclusion:
> I’m pretty sure the concept of a hidden file was an unintended consequence. It was certainly a mistake.

Regardless of the dot filename thing, since we have appropriate ways of dealing with additional options for services (e.g. `public`, `autowire`, `autoconfigure`, etc.), it only seems appropriate to also add `hidden` as a proper option to services. It doesn't make any sense to abuse the name to do this: it's the service that's hidden, not the name.

**Comment by nicolas-grekas** [comment]
The technical deciders are transitivity and simplicity: should an alias that point to a hidden service inherit this property? => No. Then how do we implement this specification? With using a name-based convention, there is nothing to do and the concern remains totally orthogonal to DI itself. Using a property, we make the concept leak in the DI world, and require a bunch of extra logic dedicated to make this work. I've been through this while working on the implementation.

**Comment by OskarStark** [comment]
Does the service (/string) need to know if its hidden or not?
if not we could do something like this:

```yaml
services:
    # default configuration for services in *this* file
    _defaults:
        autowire: true      # Automatically injects dependencies in your services.
        autoconfigure: true # Automatically registers your services as commands, event subscribers, etc.
        public: false 
    _hidden:
        my.hidden.service: ~
```

WDYT?

**Comment by nicolas-grekas** [comment]
@OskarStark there is no place to store/assign this extra semantic. We dont't have any mean to add properties on *identifiers*. And we don't want any actually, because this concept doesn't belong to DI world, but only to a listing command.

**Comment by OskarStark** [comment]
Ok, thank you for your feedback 👍

**Comment by tehplague** [comment]
+1 for adding something like `hidden: true` to the service definition. Prefixing the service identifier with a dot would break the notion of using the FQCN as service identifier as a best practice IMHO.

**Comment by tzfrs** [comment]
@nicolas-grekas 
> It's exactly the same with hidden files on Linux: this concept doe not belong to the file content nor to the file stats. It's just something that belongs to the file NAME.

But we aren't using a name of a file here. It's about a name of a service, which is different in my opinion. In CSS classes start with `.`, but they aren't hidden.

However, like file names, CSS class names are not the same as service, therefore I don't think that either could be used as an argument here. 

Also, in Windows you also use a flag on the file/folder, don't you?

**Comment by stof** [comment]
Another nice benefit of using the dot in the name, is that this forbids to have a hidden service being available to autowiring (while being hidden from all listings showing what is available).
Using an attribute would again require custom logic for that.

**Comment by syrm** [comment]
I think better than hide service could be to add category to them

```
services:
    App\Something\Foo:
        category: hidden
```

It allows to handle more case and even filter them by category in `debug:container`

**Comment by tzfrs** [comment]
Hm, not a fan of that approach tbh @syrm, because all other config values `public, autoconfigure, autowrite` simply use `true/false`, so I would keep it consistent.

**Comment by francoispluchino** [comment]
When reading the post on the blog, I also asked the question, why this is not a property `hidden` on the service? While it's much more explicit than using a dot at the beginning of a service name.

In the end, many of us say that it is much easier to understand, and therefore easier for a beginner to know that the service is hidden. It's also more consistent, because it is the same principle as the definition of commands, for example. When you start using a third-party library, the problem is not really visible (because you don't use the hidden services), but I'm thinking mostly of a beginner using a public service of the third-party library, which becomes hidden for X or Y reasons after an update. I think the beginner will not realize that the service is hidden, and cannot be used. He will look for a moment, mainly why a points was added at the beginning of the name, especially via a search engine, and as @iltar says, it will be difficult to describe the problem, if he doesn't know that it means the service is hidden.

However, the argument of @stof is relevant:

> Another nice benefit of using the dot in the name, is that this forbids to have a hidden service being available to autowiring (while being hidden from all listings showing what is available).
Using an attribute would again require custom logic for that.

I suppose that technically you use the dot to check if the service is hidden in the `debug:container` command (and elsewhere), and so, after the compilation of the cache. As a result, I suppose that the list of hidden services should be stored for cleaning, what like @nicolas-grekas indicates, represents a lot more code and complexity, whereas a simple dot in the name of the service is very easy to check (you can see the PR #26891 adding a tag to the service definitions for cleaning, and the reasons of the closing this PR in the merged PR #26921).

However, after reading everyone's arguments, I wonder, if we cannot mix the benefits of each. Namely, add the `hidden` property to the service definition, whose the compiler in her internal logic, will add a dot to the service name during the compilation. In the configuration and the doc, we can add a description on the `hidden` property indicating that a dot will be added at the beginning of the service name. In this way, we maintain consistency, simplicity, quick comprehension, while maintaining the simplicity of the current implementation.

In other words, this proposal adds simply a small configuration layer improving the DX.

Can this be considered?

**Comment by julienfalque** [comment]
A `hidden` option should not automatically add the dot in the service id: what if you want to hide the service and still have it available for autowiring? If one wants to prevent autowiring, one could still change the service id manually (with a leading dot or whatever character).

**Comment by kallosz** [comment]
if you need service available for autowiring, you shouldn't make service hidden.

**Comment by kdauzickas** [comment]
As far as I understand hiding services should not interfere in any way with how the DI actually works and this property is of interest only to the single debug command. Tagging seems like an acceptable option here. No?

**Comment by julienfalque** [comment]
@kallosz Why? In a bundle it can be useful to easily inject an internal (hidden) service into other services of the bundle.

**Comment by stof** [comment]
Well, if you hide it but keeps it available for autowiring, this will create a weird experience for users looking at what types are available for autowiring.

**Comment by Pierstoval** [comment]
Isn't it a step backwards on DX to add a special convention that developers need to know about instead of just using a very explicit feature such as `hidden: true` which is self-explanatory and does not require to read the documentation to know what it's about? I mean, I know that not many people will have to know about hidden services, but when they will, knowing that it's as simple as an option is much better than saying _"Just respect a convention and you're set"_. I thought SF was about _"Configuration over convention"_, and the "convention" part would've been only explicitly defined by the developer (like with `_defaults`)

**Comment by apfelbox** [comment]
@stof I always understood this feature as exactly that.

If I am writing a reusable bundle, I have 10 services, 2 of them should be used from the external consumer, 8 only internal. I still want to use autowiring in this bundle, because.. why not?

So I would mark 8 of these services as hidden and 2 of them not. I always thought that **exactly this usage** is the main purpose of this feature. 

That's also how I am reading the blog post, btw:

> Independently from being public or private, sometimes it's useful to hide some services from the debug:container listing because they are not intended for being used by the developers.

They are not intended to be used by developers, that are *integrating* my bundle, but I as the bundle developer should of course be able to use these services...?

**Comment by javiereguiluz** [comment]
I'm closing this issue because Nicolas said [in this comment](https://github.com/symfony/symfony/issues/27293#issuecomment-389906839):

> And there is a big misunderstanding about hidden services, #27284 should be just closed.
